### PR TITLE
Show snackbar notification when a form is saved or sent

### LINKED ISSUE #5424: Show snackbar notification when a form is saved or sent
**Acceptance**
- [x] Given I'm filling or editing a form
And I'm at the end of the form
When I click "Save as Draft"
Then I am returned to the Main Menu
And I see a snack bar reading "Your form was saved as a draft." with an "Edit" action and a dismiss (X) button

- [x] Given I'm filling or editing a form
And I'm at the end of the form
And I click "Save as Draft"
And I click "Edit" on the snack bar
Then the form instance is opened for editing

- [x] Given I'm filling or editing a form
And I'm at the end of the form
When I click "Finalize"
Then I am returned to the Main Menu
And I see a snack bar reading "Your form was saved." with a "View" action and a dismiss (X) button

- [x] Given I'm filling or editing a form
And I'm at the end of the form
And I click "Finalize"
And I click "View" on the snack bar
Then the form instance is opened in read-only mode

- [x] Given I'm filling or editing a form
And auto send is enabled
And the device is connected to the internet
And I'm at the end of the form
When I click "Send"
Then I am returned to the Main Menu
And I see a snack bar reading "Sending form..." with a "View" action and a dismiss (X) button

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5424 

#### What has been done to verify that this works as intended?
I've tested the changes manually and added new automated tests.

#### Why is this the best possible solution? Were any other approaches considered?
The most important thing here is how I handled passing an URI of a saved form to the main menu in order to display a snackbar with appropriate data. Initially I did it using shared prefs https://github.com/getodk/collect/pull/5609/commits/4e515a5b63ba1853294490b28c43ec62dbcb47f6 but then decided to take advantage of the fact that when a form is saved the URI is passed using activity result. This required some changes https://github.com/getodk/collect/pull/5609/commits/0bfc7acbd9011acd247525d45bc1edc2f145b2b3 in the way we start the list of forms and the list of drafts to handle the result but I think this solution is more elegant and consistent.

Another thing of note is that I removed marking forms as failed before sending them see https://github.com/getodk/collect/pull/5609/commits/227adcd7e70e2ff7a3b6cc81724fe948b1b3109a
This was implemented to avoid editing forms that should be sent but now we do not allow editing finalized forms at all so it is redundant.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Please review the acceptance criteria in the issue and test saving forms according to them. When a snackbar is being created we assume that a form is saved as draft or finalized (and should be sent or not) but it's not sent yet. It would be good to check if it is possible for example using fast devices and small forms that the form is sent (or failed to send) before the process of creating a snackbar is finished because in such a case it won't be displayed at all.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Review by seadowg — CHANGES_REQUESTED** [review]
Now we've got this to play with, I think there are a couple of behavioural things we should change before diving into a code review:

* The snackbar should be hidden after an action (like "View") is clicked. Right now it's on-screen for a while (which is good), but if you click "View" and then press back to return the menu, the snack bar is still there and is confusing.
* The "It looks like you're offline..." doesn't really make sense as-is. We only show it if the device is connected to the network required for auto send, but somehow also offline which will happen very rarely (even if at all) in reality. I think we should show also show this message in the cases that auto send won't happen (auto send is wifi only, but we only have a cellular connection for instance). We might need to tweak the text a little as "offline" won't be strictly true any longer. It'd be good to get some thoughts from @alyblenkin on this one.

**Comment by alyblenkin** [comment]
> Now we've got this to play with, I think there are a couple of behavioural things we should change before diving into a code review:
> 
> * The snackbar should be hidden after an action (like "View") is clicked. Right now it's on-screen for a while (which is good), but if you click "View" and then press back to return the menu, the snack bar is still there and is confusing.

Ah yes, it should go away if you press back to return to the main menu. You wouldn't want to see the snackbar after you've take an action, it should disappear. 

> * The "It looks like you're offline..." doesn't really make sense as-is. We only show it if the device is connected to the network required for auto send, but somehow also offline which will happen very rarely (even if at all) in reality. I think we should show also show this message in the cases that auto send won't happen (auto send is wifi only, but we only have a cellular connection for instance). We might need to tweak the text a little as "offline" won't be strictly true any longer. It'd be good to get some thoughts from @alyblenkin on this one.

I didn't think about this - good point! 

If the device is connected to the network required for auto-send, we show "It looks like you're offline..." 

If they have a cellular connection, but no wifi, perhaps we could show "It looks like you don't have Wi-Fi. Your form will send when you are online" or "connected"? I guess they technically have a cellular connection. "It looks like you don't have Wi-Fi. Your form will be sent when you connect" 

What do you think?

**Comment by grzesiek2010** [comment]
> If they have a cellular connection, but no wifi, perhaps we could show "It looks like you don't have Wi-Fi. Your form will send when you are online" or "connected"? I guess they technically have a cellular connection. "It looks like you don't have Wi-Fi. Your form will be sent when you connect"

Now I wonder if it make sense to check internet connection and display it in the snackbar at all...
1. Theoritecally it is possible that there is no connection when the snackbar is being created but 1s later when the process of sending forms starts the devices is connected or the other way around so the message might not be accurate.
2. We would need to check more cases and probably display different messages:
   - `wifi only` is selected in settings but you have no connection or only cellular 
   - `cellular only` is selected in settings but you have only wifi
   - `wifi or cellular` is selected in settings but there is no connection at all
3. When auto-send fails we display a notification and I think that notification should contain a proper message that explain what happened.

Taking into account all of that maybe the snackbar should always display `Sending form…` if autosend is enabled and that notification should be responsible for providing more info.

**Comment by seadowg** [comment]
EDIT: I wrote this before I saw @grzesiek2010's reply (was offline), so some of the comments are redundant. Sorry!

> If the device is connected to the network required for auto-send, we show "It looks like you're offline..."

Do you mean if they're connected, but their connection isn't working (like bad signal for instance)? We won't be able to detect that scenario dependably. The only times the phone really knows it's "offline" it can't find a cellular or Wi-Fi network at all. We could try and determine whether we have a "good" connection or not (pinging a known dependable server for instance), but I think that's further than we want to go here (and it's not trivial to do if we're needing to determine the quality of a specific connection type).

> If they have a cellular connection, but no wifi, perhaps we could show "It looks like you don't have Wi-Fi. Your form will send when you are online" or "connected"? I guess they technically have a cellular connection. "It looks like you don't have Wi-Fi. Your form will be sent when you connect"

I think this approach makes sense. I guess the different scenarios when auto send is enabled we need to cover could look like this:

1. wifi and cellular enabled, but neither available
2. wifi only, but not available (cellular may or may not be enabled but doesn't change anything)
3. cellular only, but not available (wifi may or may not be enabled but doesn't change anything)

As I'm writing this I'm realizing there's further complexity here: in situation 2 or 3, I don't think we'll ever auto send the form if the connection type we need is disconnected, but the other is working. This is because (as far as I can tell) we currently wait for any connection to try and auto send, but will then determine whether we have the correct connection type or not and (again, as far as I can see), won't reschedule trying to send again if we don't. This would mean in a case where we have a cellular connection, but no wifi and auto send is "wifi only", we'd actually never auto send the form, so we can't say  "Your form will be sent when you connect".

@grzesiek2010 could you investigate this to see if I'm correct about those scenarios or not? I think we might want to either rethink what we say on this snackbar, or try and improve the auto send behaviour before adding it (which I'd prefer).

**Comment by grzesiek2010** [comment]
> This would mean in a case where we have a cellular connection, but no wifi and auto send is "wifi only", we'd actually never auto send the form, so we can't say "Your form will be sent when you connect".

In such cases, the `Send` button in the end screen is not displayed instead we see the `Finalize` one so it wouldn't be a problem. The message in snackabr would also be `Your form was saved` and not `Sending form...`.

But again the more I think about it the more I'm convinced that if the button is `Send` we should always display just `Sending form...` because as I said above at the time when the scankbar is being created the connection might not be available but after 1s it changes and the form is sent or the other way around. This solution would be simpler and more accurate and potential errors should be displayed in a notification.

**Comment by seadowg** [comment]
> But again the more I think about it the more I'm convinced that if the button is Send we should always display just Sending form... because as I said above at the time when the scankbar is being created the connection might not be available but after 1s it changes and the form is sent or the other way around

Yeah I agree with this. I think we should go with this for the moment (not have a different message for "offline"), and then look at revising auto send to be less brittle in the future. I think there's a lot we could improve around retrying to send when we have the correct connection that would allow us to more accurately say to the user "Your form will be sent later". Does that sound ok @alyblenkin?

If so, let's remove that offline message and hide the snackbar when an action is clicked as discussed [here](https://github.com/getodk/collect/pull/5609#pullrequestreview-1447384244).

**Comment by lognaturel** [comment]
Agreed just giving feedback that a send attempt is ongoing and giving a way to immediately view the form that was just completed is already very helpful. Seems reasonable to focus on that for this release.

Have we also considered showing a snackbar when an autosend attempt completes? Currently we show a notification but my experience is that those are really easy to miss and they feel less tied to current actions. I think we may have discussed switching to or adding snackbars at some point.

For example:
- autosend runs, a single form is successfully sent: "Form sent. &nbsp;&nbsp;&nbsp;&nbsp;View"
- autosend runs, multiple forms are successfully sent: "N forms sent."
- autosend runs, there are failures: "Autosend failure&nbsp;&nbsp;&nbsp;&nbsp;Go to Ready to send" (not sure what that action link should be)

**Comment by alyblenkin** [comment]
> Agreed just giving feedback that a send attempt is ongoing and giving a way to immediately view the form that was just completed is already very helpful. Seems reasonable to focus on that for this release.
> 
> Have we also considered showing a snackbar when an autosend attempt completes? Currently we show a notification but my experience is that those are really easy to miss and they feel less tied to current actions. I think we may have discussed switching to or adding snackbars at some point.
> 
> For example:
> 
> * autosend runs, a single form is successfully sent: "Form sent.     View"
> * autosend runs, multiple forms are successfully sent: "N forms sent."

I spoke to @seadowg about the successfully sent message a while back, and we were worried that they could appear out of context or mid-flow when they are filling out another form. I documented some of that [conversation here](https://www.notion.so/System-notifications-e29d1836cdf744d6905794cebb312b25) - @seadowg suggested we could maybe leverage system notifications down the road.

> * autosend runs, there are failures: "Autosend failure    Go to Ready to send" (not sure what that action link should be)


What about - "Unable to send. Your form is "Ready to Send" later.  View"

**Comment by grzesiek2010** [comment]
> If so, let's remove that offline message and hide the snackbar when an action is clicked as discussed

Done.

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java [hunk 1]
```diff
@@ -72,6 +74,11 @@ public class InstanceChooserList extends InstanceListActivity implements Adapter
     @Inject
     FormsRepositoryProvider formsRepositoryProvider;
 
+    private final ActivityResultLauncher<Intent> formLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
+        setResult(RESULT_OK, result.getData());
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java:83]: This isn't covered by any tests as far as I can tell.

File: collect_app/src/main/java/org/odk/collect/android/upload/InstanceServerUploader.java [hunk 2]
```diff
@@ -71,8 +71,6 @@ public InstanceServerUploader(OpenRosaHttpInterface httpInterface,
      */
     @Override
     public String uploadOneSubmission(Instance instance, String urlString) throws FormUploadException {
-        markSubmissionFailed(instance);
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/upload/InstanceServerUploader.java:74]: Is there any benefit to changing this?

File: testshared/build.gradle [hunk 3]
```diff
@@ -51,6 +51,7 @@ dependencies {
     implementation Dependencies.androidx_test_espresso_contrib
     implementation Dependencies.android_material
     implementation Dependencies.danlew_android_joda
+    implementation Dependencies.mockito_kotlin
```
Comment (seadowg) [inline testshared/build.gradle]: I don't think you ended up needing this

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormSavedSnackbarTest.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.RuleChain
+import org.odk.collect.android.R
+import org.odk.collect.android.support.pages.MainMenuPage
+import org.odk.collect.android.support.pages.ProjectSettingsPage
+import org.odk.collect.android.support.rules.CollectTestRule
+import org.odk.collect.android.support.rules.TestRuleChain
+
+class FormSavedSnackbarTest {
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormSavedSnackbarTest.kt:11]: I think you only need to test the saving a new form and saving a draft cases here (to check that the activity result setup is correct, and the Activity is hooked up to the ViewModel). Most of the details of the snack bar seem to be covered in the tests for `MainMenuViewModel#getFormSavedSnackbarType` in `MainMenuViewModelTest`.

File: collect_app/src/main/java/org/odk/collect/android/mainmenu/MainMenuViewModel.kt [hunk 5]
```diff
@@ -95,20 +105,45 @@ class MainMenuViewModel(
     val sentInstancesCount: LiveData<Int>
         get() = instancesAppState.sentCount
 
+    fun getFormSavedSnackbarType(uri: Uri): FormSavedSnackbarType? {
+        val instance = instancesRepositoryProvider.get().get(ContentUriHelper.getIdFromUri(uri))
+        if (instance != null) {
+            return when (instance.status) {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/mainmenu/MainMenuViewModel.kt]: This `return` can be lifted to the `if`

**Review by seadowg — CHANGES_REQUESTED** [review]
Looking good! Just a few things inline, but nothing major.

Before merging, it would be good to rebase this on top of `master` so that it gets the latest fixes for our instrumentation tests as this is red on Test Lab right now.

File: collect_app/src/main/java/org/odk/collect/android/upload/InstanceServerUploader.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/upload/InstanceServerUploader.java:74]: 1. It was a confusing workaround. If the form was marked a failed before sending it was also visible to a user.
2. When creating a snackbar I need a real status but here it was possible that a form would be marked as failed first.

File: collect_app/src/main/java/org/odk/collect/android/mainmenu/MainMenuViewModel.kt (hunk 5 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/mainmenu/MainMenuViewModel.kt:111]: You mean something like this:
```java
    fun getFormSavedSnackbarType(uri: Uri): FormSavedSnackbarType? {
        val instance = instancesRepositoryProvider.get().get(ContentUriHelper.getIdFromUri(uri))
        if (instance != null) {
            if (instance.status == Instance.STATUS_INCOMPLETE) {
                return FormSavedSnackbarType.SAVED_AS_DRAFT
            } else if (instance.status == Instance.STATUS_COMPLETE) {
                val form = formsRepositoryProvider.get().getAllByFormIdAndVersion(instance.formId, instance.formVersion).first()
                return if (form.shouldFormBeSentAutomatically(autoSendSettingsProvider.isAutoSendEnabledInSettings())) {
                    FormSavedSnackbarType.SENDING
                } else {
                    FormSavedSnackbarType.FINALIZED
                }
            }
        }
        return null
    }
```

I don't see any advantage of this change.

File: testshared/build.gradle (hunk 3 above)
Comment (grzesiek2010) [reply testshared/build.gradle]: Right.

File: collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/InstanceChooserList.java:83]: Fixed in https://github.com/getodk/collect/pull/5609/commits/9930c96c2c983c7f2a1654cc1a36d160ebfb4955

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormSavedSnackbarTest.kt (hunk 4 above)
Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormSavedSnackbarTest.kt:11]: The tests in MainMenuViewModelTest verify that a proper SnackbarType is returned based on form status but they do not verify:
- if the snackbar is displayed at all
- if it contains proper message and action
- if a form is opened for edit/view after clicking on the action

File: collect_app/src/main/java/org/odk/collect/android/mainmenu/MainMenuViewModel.kt (hunk 5 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/mainmenu/MainMenuViewModel.kt]: Done.

File: collect_app/src/main/java/org/odk/collect/android/upload/InstanceServerUploader.java (hunk 2 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/upload/InstanceServerUploader.java:74]: Thanks that all makes sense!

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormSavedSnackbarTest.kt (hunk 4 above)
Comment (seadowg) [reply collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormSavedSnackbarTest.kt:11]: As far as I can see, there is no conditional logic in `displayFormSavedSnackbar`. Therefore, if you just had a test that checks the snack bar flow for a new form (as finalized or a draft) and another to check the flow when editing and then saving a draft (as finalized or a draft), that'd check that the code that shows the snack bar is there and that the activity result stuff (why you need two tests for the two Activity objects) is all hooked up correctly. Then, the other cases are all covered by your tests in `MainMenuViewModelTest` as you check there that you get the right messages and actions in the right scenarios. As far as I understand whether the form gets opened for editing/viewing etc is handled by `FormUriActivity` and that is all tested in `FormUriActivityTest`. 

Basically, I'm imagining the top level of tests (instrumented) just check "we show a snack bar after filling a new form" and "we show a snack bar after editing a draft" and that the ViewModel tests and Activity tests (local) can cover the details like how the form was saved, is auto send on etc.

Does that make sense? Happy to chat about it more!

**Comment by seadowg** [comment]
I'm going to mark this "needs testing", as we're only discussing tests now.

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormSavedSnackbarTest.kt (hunk 4 above)
Comment (grzesiek2010) [reply collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormSavedSnackbarTest.kt:11]: Ah yeah it makes sense.

**Review by seadowg — APPROVED** [review]

**Comment by dbemke** [comment]
I'm not sure what should be an expected result in some scenarios:
1.   Steps:
-  Go to Drafts
- Open a form, go to a question
- Tap the device back button and **discard** changes

In this case the user is moved to the main menu and sees "Your form was **saved** as a draft" 

2. If there are few forms in Ready to send that are being sent at the moment and I'm on the main menu should I see the snackbar "Sending form..." for each form when it's being sent?

**Comment by dbemke** [comment]
3. Steps:
-  In Form Entry settings disable moving backwards.
- Go to Start new form and open a form.
- Fill the form and save it as a draft
- When the snackbar appears click "Edit”

The user is moved to the view of a form with only an "Exit" option which seems ok but on the other hand moving backwards is supposed to prevent edits(?) so the word Edit in snackbar might be confusing?

**Comment by dbemke** [comment]
Similar scenario to the previous one
4. Steps:
- In Access control, Main Menu Settings disable Drafts
- Go to Start new form and open a form.
- Fill the form and save it as a draft
- When the snackbar appears click "Edit”

The user is moved to the view of a form with only an "Exit" option

**Comment by grzesiek2010** [comment]
> Go to Drafts
Open a form, go to a question
Tap the device back button and discard changes
In this case the user is moved to the main menu and sees "Your form was saved as a draft"

Not perfect but also not something terrible to me. It happens because the snackbar is created whenever after quitting a form its URI is returned. It turns out that when we close a form discarding changes it is returned. I'm not sure if it's the expected behavior but we probably shouldn't touch it so fi we want to improve the current behavior we need to add something to the returned intent when the form is closed to distinguish those cases.

> If there are few forms in Ready to send that are being sent at the moment and I'm on the main menu should I see the snackbar "Sending form..." for each form when it's being sent?

No, the snackbar should be displayed only after closing a form for that particular form.

> In Form Entry settings disable moving backwards.
Go to Start new form and open a form.
Fill the form and save it as a draft
When the snackbar appears click "Edit”

Here the action name should be `View` instead of `Edit`

> In Access control, Main Menu Settings disable Drafts
Go to Start new form and open a form.
Fill the form and save it as a draft
When the snackbar appears click "Edit”

As above the action name should be `View` instead of `Edit`

I also wonder if it makes sense to have that `Save as draft` options visible if editing drafts is disabled like in those two last cases but it's something for a separate discussion I guess.

@seadowg what do you think about all those cases? Do you agree with what I said?

**Comment by dbemke** [comment]
One more question about expected
5. Steps:
- In settings, access control, main menu settings disable Ready to send and Sent
- Go to a form and send it

The users sees a snackbar with "View" option. I don't know if hiding Ready to send and Sent option from the main menu means that the user is not supposed to view forms that are being sent/ were sent?

**Comment by grzesiek2010** [comment]
> In settings, access control, main menu settings disable Ready to send and Sent
Go to a form and send it
The users sees a snackbar with "View" option. I don't know if hiding Ready to send and Sent option from the main menu means that the user is not supposed to view forms that are being sent/ were sent?

Hmm If viewing sent forms was only in the list of sent forms (not in the list of forms to send too) I would say that the View option in the snackbar should not be available but now I don't know... I would file a separate issue for this and discuss it later.

**Comment by seadowg** [comment]
For **1**, I think we should fix this now by not showing a snackbar when exiting a form via "Discard changes/form". In terms of implementation, I think it should be fairly feasible to just not set a result (or set a `RESULT_CANCELLED` result) in `FormFillingActivity` when changes are discarded so that the activity result callbacks never end up firing. If it ends up being much more complex than that, then I agree it's something we should look at as a separate PR.

Totally agree that for cases **3** and **4** we should just change the action text from "Edit" to "View". It might be that there's a helper we can pull out (probably from the logic in `FormUriActivity`) so that we don't end up duplicating things.

I also agree that **5** is something we should discuss as part of another issue - we might not make any changes there.

**Comment by grzesiek2010** [comment]
> In terms of implementation, I think it should be fairly feasible to just not set a result (or set a RESULT_CANCELLED result) in FormFillingActivity when changes are discarded so that the activity result callbacks never end up firing.

You mean removing returning URI in that case like in https://github.com/getodk/collect/pull/5609/commits/76ec768a0f7b713cb34bec1b7672190b290796c0
Generally, I was thinking about that, and to me, that didn't even make sense (retuning that uri) but still, I'm not sure if it's not going to break anything...

**Comment by seadowg** [comment]
@grzesiek2010 Yes that makes sense to me. It's a change in behaviour that will affect external apps (calling into Collect), so we'll need to announce the change explicitly, but that's ok.

**Comment by srujner** [comment]
@dbemke already tested this PR

**Tested With Success!**

**Verified on Device with Android**: 10, 12, 13

**Verified Cases**:
- Acceptance Criteria from #5424 has been met;
- Issues mentioned by Dominika has been fixed;
- Number **5** will be created as separate Issue to discuss;
- Snackbars with all auto send options;
- Snackbars while auto sent is set to true or false and auto delete is set to true or false in a form;
- Switching to airplane mode while on the end page and trying to send a form in Start new form and Drafts with saving as draft enabled and disabled in Form End;
- Checking audit logs after editing and viewing the form via snackbar in Track changes and Track changes reason, Identify user forms;
- Snackbars while moving backwards is disabled, Drafts disabled in the main menu;
- Using Save as Draft option when Exiting the Form;
- Viewing a sent form via snackbars while in settings Delete after send is enabled and disabled;
- Sending a form while in server setting there is a wrong URL;
- Different font sizes;
- Don’t keep activities enabled/disabled;
- In device settings - no background processes;
- Saving, sending a form added from the form map view;
- Light and Dark Mode;
- Rotate the screen, minimize the app.
